src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs

```diff
@@ -270,20 +277,52 @@ public WeakReference(ReferenceCountedDisposable<T> reference)
             /// already been disposed.</returns>
             public ReferenceCountedDisposable<T>? TryAddReference()
             {
-                var weakInstance = _weakInstance;
-                if (weakInstance == null || !weakInstance.TryGetTarget(out var target))
+                // Ensure 'this' is only read once by this method.
+                var referenceCount = _boxedReferenceCount;
+                if (referenceCount == null)
                 {
+                    // This is either a default(WeakReference), or the current instance was constructed from a reference
+                    // that was already disposed. No target is available.
                     return null;
                 }
 
-                var referenceCount = _boxedReferenceCount;
-                if (referenceCount == null)
+                var weakInstance = referenceCount._weakInstance;
+
+                // _weakInstance is initialized by the constructor before assigning a value to _boxedReferenceCount.
+                // Since it latches in a non-null state, it cannot be null at this point.
+                Contract.ThrowIfNull(weakInstance);
+
+                if (!weakInstance.TryGetTarget(out var target))
                 {
+                    // The weak reference has already been collected, so the target is no longer available.
                     return null;
                 }
 
                 return TryAddReferenceImpl(target, referenceCount);
             }
         }
+
+        /// <summary>
+        /// Holds the reference count associated with a disposable object.
+        /// </summary>
+        private sealed class BoxedReferenceCount
+        {
+            /// <summary>
+            /// Holds the weak reference used by instances of <see cref="WeakReference"/> to obtain a reference-counted
+            /// reference to the original object. This field is initialized the first time a weak reference is obtained
+            /// for the instance, and latches in a non-null state once initialized.
```

Comment:
Can we apply the [DisallowNull] or whatever it is to make the latching behavior clear?

Reply:
➡️ Applied attribute